### PR TITLE
Calling prepare_model_for_kbit_training if you aren't quantizing the model can freeze all parameters when doing LoRA training

### PR DESCRIPTION
**Pull request by njbrake**
Hi! I am still trying to track down the exact failure condition, but it seems that if you aren't 8bit or 4bit quantizing the model, then calling prepare_model_for_kbit_training freezes all the weights in the model and training will fail with the error

`warnings.warn("None of the inputs have requires_grad=True. Gradients will be None")`

```
model = AutoModelForCausalLM.from_pretrained(
        model_args.model_name_or_path,
        load_in_8bit=False,
        load_in_4bit=False,
        device_map=device_map,
    )
model = prepare_model_for_kbit_training(model)

trainable_params = [p for p in model.parameters() if p.requires_grad]
if len(trainable_params) == 0:
    print("No trainable parameters in the model.")

config = LoraConfig(
    r=model_args.lora_r,
    lora_alpha=model_args.lora_alpha,
    target_modules=model_args.lora_target_modules,
    lora_dropout= model_args.lora_dropout,
    bias="none",
    task_type="CAUSAL_LM",
)
model = get_peft_model(model, config)
for n, p in model.named_parameters():
    print(f"{n}: {p.requires_grad}")
```

Seems like there should at least be a warning message if you call prepare_model_for_kbit_training but didn't set `load_in_8bit` or `load_in_4bit` to True?

### CODE REVIEW

**Comment by njbrake** [comment]
What is confusing me is that for the above settings, that second statement shows 
```
base_model.model.model.layers.18.self_attn.q_proj.weight: False
base_model.model.model.layers.18.self_attn.q_proj.lora_A.default.weight: True
base_model.model.model.layers.18.self_attn.q_proj.lora_B.default.weight: True
base_model.model.model.layers.18.self_attn.k_proj.weight: False
```
But yet, training is still failing with the message `None of the inputs have requires_grad=True. Gradients will be None`.

Is this related to the code 
```
        # For backward compatibility
        if hasattr(model, "enable_input_require_grads"):
            model.enable_input_require_grads()
```
in the prepare_model_for_kbit_training function? Seems suspect as both are mentioning `input`.

File: src/peft/utils/other.py [hunk 1]
```diff
@@ -72,7 +72,11 @@ def prepare_model_for_kbit_training(model, use_gradient_checkpointing=True):
             The loaded model from `transformers`
     """
     loaded_in_kbit = getattr(model, "is_loaded_in_8bit", False) or getattr(model, "is_loaded_in_4bit", False)
-
+    if not loaded_in_kbit:
+        warnings.warn(
+            "You are not quantizing the model, This function is going to freeze all of the base model's parameters. If you are going to train a new LoRA adapter, there may be no parameters with requires_grad=True.",
```
Comment (BenjaminBossan) [inline src/peft/utils/other.py:77]: This line is too long, could you introduce some line breaks (120 width)?

File: src/peft/utils/other.py [hunk 2]
```diff
@@ -72,7 +72,11 @@ def prepare_model_for_kbit_training(model, use_gradient_checkpointing=True):
...
+            Warning
```
Comment (BenjaminBossan) [inline src/peft/utils/other.py:78]: I don't think we need the `Warning` argument.

File: src/peft/utils/other.py [hunk 3]
```diff
@@ -95,7 +99,13 @@ def make_inputs_require_grad(module, input, output):
 
         # enable gradient checkpointing for memory efficiency
         model.gradient_checkpointing_enable()
-
+    else:
+         # check to see if the model has any parameters that are trainable
+        if len([p for p in model.parameters() if p.requires_grad]) == 0:
+            warnings.warn(
+                "This model has no trainable parameters: did you accidentally call prepare_model_for_kbit_training when your model was not quantized?",
+                Warning
```
Comment (BenjaminBossan) [inline src/peft/utils/other.py:107]: I don't think we need the `Warning` argument.

**Review by BenjaminBossan — CHANGES_REQUESTED** [review]
Thanks for the PR.

I think we don't we need both warnings, IIUC, for the situation you describe, both warnings would be given? Maybe the 2nd warning is the right one to keep.

**Review by younesbelkada — APPROVED** [review]
Looks great, thanks!

**Comment by pacman100** [comment]
Hello @njbrake, which version of PEFT are you using?

**Comment by pacman100** [comment]
I don't think this PR is needed. In `prepare_model_for_kbit_training`, we specifically make all the params to not require grads:
```
for name, param in model.named_parameters():
        # freeze base model's layers
        param.requires_grad = False
```

So, the check to see if there are any trainable params and the following warning is redundant

**Comment by njbrake** [comment]
@pacman100 You make a great point, no need to check to see if there are any trainable params, since we had frozen them just a few lines above. 

Thanks for the feedback! I found the issue. I'm trying to figure out the best way to address it. As you saw, I have gradient_checkpointing enabled, which is a part of the Huggingface Training Args. https://github.com/huggingface/transformers/blob/main/src/transformers/training_args.py#L570

However, The peft model is loaded before the trainer is run, and since the check that you reference via [code](https://github.com/huggingface/peft/blob/main/src/peft/peft_model.py#L120-L121)  happens in the `__init__` of the PEFT model,  this logic `getattr(model, "is_gradient_checkpointing", True)` is going to evaluate as false because it is the Trainer that is enabling gradient_checkpointing of the model, and the Trainer hasn't yet run.

So, now I think about where we should go from here. 

I could of course add the following to my code before loading the PEFT model on top of the AutoModelForCausalLM

```
if training_args.gradient_checkpointing:
    model.gradient_checkpointing_enable()
```

But this seems like an undesirable "gotcha" because isn't the point of the Trainer that it should take care of this for me? Or maybe we should add some sort of logging into the Trainer to check if the model is a PEFT model and if gradient_checkpointing was enabled via the training args but not yet set in the model, we throw a warning to the user to tell them of the issue.

Thanks!

**Comment by njbrake** [comment]
Hi @pacman100, Checking in to see if you had any thoughts about my comment above?
